File: tags/ARTServer_130521/src/graphics/Scene.java
package graphics;

import graphics.drawables.IDrawable;
import graphics.drawables.ImageDrawable;

import java.awt.Color;
import java.util.Hashtable;
import java.util.LinkedList;


public final class Scene
{
	public static final String LAYER_NAME_BACKGROUND	= "b";
	public static final String LAYER_NAME_FOREGROUND	= "f";
	public static final String LAYER_NAME_OVERLAY		= "o";
	public static final String LAYER_NAME_WIDGETS		= "w";
	
	private boolean _isDrawing = false;
	private final double _width;
	public double getWidth() { return _width; }
	
	private final double _height;
	public double getHeight() { return _height; }

	private final double _scale;
	
	public final class Layer
	{
		public IDrawable clipMask = null;
		private LinkedList<IDrawable> _drawables = new LinkedList<IDrawable>();
		public boolean add(IDrawable drawable)
		{
			_drawables.add(drawable);
			return true;
		}
		
		public boolean remove(IDrawable drawable)
		{
			return _drawables.remove(drawable);
		}
		
		void onDraw(SceneContext params)
		{
	        params.context.save();
	        if(this.clipMask != null) this.clipMask.onDraw(params);

	        for(IDrawable drawable : _drawables)
	        {
	        	if(drawable instanceof ImageDrawable)
	        	{
	        		drawable.onDraw(params);
	        	}
	        }

	        for(IDrawable drawable : _drawables)
	        {
	        	if(!(drawable instanceof ImageDrawable))
	        	{
	        		drawable.onDraw(params);
	        	}
	        }
	        params.context.restore();
		}
	}
	
	private Hashtable<String, Layer> _layers = new Hashtable<String, Layer>();
	private LinkedList<Layer> _zorder = new LinkedList<Layer>();
	
	public Layer addLayer(String name)
	{
        Layer newLayer = new Layer(); 
        _zorder.addLast(newLayer);
        _layers.put(name, newLayer);
        return  newLayer;
	}
	
	public Layer getLayer(String name)
	{
		return _layers.get(name);
	}
	
	public void drawTo(JavaCanvas canvas, Color inkColor)
	{
		if(_isDrawing) return;

		_isDrawing = true;
		double offsetX = _width / 2.0;
		double offsetY = _width / 2.0;

		SceneContext params = new SceneContext(
				new Rect2D( -offsetX, -offsetY, offsetX, offsetY ), 
    			canvas
    		);
		params.palette[SceneContext.PALETTE_COLOR_INK] = inkColor;
		
        params.context.save();

		try
    	{
	        
			params.context.translate(canvas.getWidth() / 2.0, canvas.getHeight() / 2.0);
					
			params.context.scale(_scale, _scale);
			
	        for(Layer layer : _zorder)
	        {
	        	layer.onDraw(params);
	        }
    	}
    	finally
    	{
        	params.context.restore();
        	_isDrawing = false;
    	}
	}
	
	public Scene(double width, double height, double scale)
	{
		_width = width;
		_height = height;
		_scale = scale;
	}
}

File: tags/120906/js/design/scene.js
function RectangularClipMask(x1, y1, x2, y2)
{
    this.x1 = x1;
    this.y1 = y1;
    this.x2 = x2;
    this.y2 = y2;

    this.onDraw = function(params)
    {
        ctx.beginPath();
        //ctx.arc(0,0,60,0,Math.PI*2,true);
        ctx.clip();
    };
}

function CircularClipMask(x, y, radius)
{
    this.x = x;
    this.y = y;
    this.radius = radius;

    this.onDraw = function(params)
    {
        var ctx = params.context;
        ctx.beginPath();
        ctx.arc(
            this.x, this.y,
            this.radius,
            0,Math.PI*2,false
        );
        ctx.fillStyle = params.palette.paper;
        ctx.fill();
        ctx.clip();
    };
}

function SceneLayer(scene)
{
    var drawables = [];
    
    
    this.visible = true;
    this.clipMask = null;
    
    this.add = function(drawable)
    {
        drawables.push(drawable);
        return true;            
    };
    
    this.remove = function(drawable)
    {
        for(var i in drawables)
        {
            if(drawables[i] === drawable)
            {                
                if(drawables.onRelease)
                {
                	drawables.onRelease(scene.getCommonParams());
                }
                drawables.splice(i,1);
                return true;
            }           
        }
        return false;           
    };
    
    this.onDraw = function(params)
    {
        params.context.save();
        if(this.clipMask != null) this.clipMask.onDraw(params);

        for(var i=0; i<drawables.length; i++)
        {
            drawables[i].onDraw(params);
        }
                
        params.context.restore();
    };  

    this.onPress = function(params)
    {
        for(var i = drawables.length-1; i>=0; i--) 
        {
            if(drawables[i].onPress)
            {
                var capture = drawables[i].onPress(params); 
                if(capture) return capture;             
            }
        }
        return null;
    };

    this.onRelease = function(params)
    {
        for(var i in drawables)
        {
            if(drawables[i].onRelease)
            {
                var capture = drawables[i].onRelease(params); 
                if(capture) return capture;             
            }
        }
        return null;
    };

    this.onMove = function(params)
    {
        for(var i in drawables)
        {
            if(drawables[i].onMove)
            {
                var capture = drawables[i].onMove(params); 
                if(capture) return capture;             
            }
        }
        return null;
    };
}


function Scene(canvasId, width, height)
{
    var layers = {};
    var zorder = [];
    var capture = null;
    var context = null;
    
	this.inkColor = "black";//"#000000";
    
    

    this.getCommonParams = function()
    {
        return {
            dirty : false,
            scale : this.scale,
            palette : {
            	ink: this.inkColor,
            	paper: "#ffffff",
            	background: "#404040",
            	highlight:"rgba(180,240,255,0.75)"
            },
            viewRect : {
                x1 : -this.offsetX,
                y1 : -this.offsetY,
                x2 : -this.offsetX + this.width / this.scale,               
                y2 : -this.offsetY + this.height / this.scale               
            }
        };
    }

    this.canvasIs = canvasId;
    this.scale = 1.0;
    this.offsetX = width / 2;
    this.offsetY = height / 2;
    this.width = width;
    this.height = height;
    
    this.addLayer = function(name)
    {
        var newLayer = new SceneLayer(this); 
        zorder.push(newLayer);
        layers[name] = newLayer;
        return  newLayer;
    };
    
    this.getLayer = function(name)
    {
        return layers[name];
    };
    
    this.onPress = function(x, y)
    {
        var params = this.getCommonParams();
        params.x = (x - this.offsetX) / this.scale;
        params.y = (y - this.offsetY) / this.scale;
        
        for(params.pass=0; params.pass<2; params.pass++)
        {
            //for(var i = 0; i < zorder.length; i++) 
            for(var i = zorder.length-1; i>= 0; i--) 
            {
                capture = zorder[i].onPress(params); 
                if(capture) break;              
            }
            if(capture != null) break;
        }

        if(params.dirty) this.redraw();
        return capture; 
    };

    this.onRelease = function(x, y)
    {
        var params = this.getCommonParams();
        params.x = (x - this.offsetX) / this.scale;
        params.y = (y - this.offsetY) / this.scale;
        
        if(capture)
        {
            capture.onRelease(params);
        }
        else
        {
            for(var i = zorder.length-1; i>= 0; i--) 
            {
                if(zorder[i].onRelease(params)) break; 
            }
        }
        capture = null;

        if(params.dirty) this.redraw(); 
    };

    this.onMove = function(x, y)
    {
        var params = this.getCommonParams();
        params.x = (x - this.offsetX) / this.scale;
        params.y = (y - this.offsetY) / this.scale;
        
        if(capture)
        {
            capture.onMove(params);
        }
        else
        {
            for(var i = zorder.length-1; i>= 0; i--) 
            {
                if(zorder[i].onMove(params)) break; 
            }
        }

        if(params.dirty) this.redraw(); 
        return capture; 
    };

    this.redraw = function()
    {
        var canvas = document.getElementById(this.canvasIs);        
        var params = this.getCommonParams();
        
        params.context = canvas.getContext("2d");
        params.context.clearRect( 0 , 0 , canvas.width , canvas.height );        
        params.context.save();
		params.context.translate(this.offsetX, this.offsetY);		        
        for(var i=0; i<zorder.length; i++)
        {
            if(zorder[i].visible)
            {
                zorder[i].onDraw(params);
            }
        }
        params.context.restore();
    };

    this.addLayer(Scene.LAYER_BACKGROUND);
    this.addLayer(Scene.LAYER_FOREGROUND);
    this.addLayer(Scene.LAYER_OVERLAY);
    this.addLayer(Scene.LAYER_WIDGETS);
}

Scene.LAYER_BACKGROUND  = "b";
Scene.LAYER_FOREGROUND  = "f";
Scene.LAYER_OVERLAY     = "o";
Scene.LAYER_WIDGETS     = "w";